src/group_theory/finite_abelian.lean

```diff
@@ -17,9 +17,29 @@ import algebra.module.pid
 
 -/
 
-
 open_locale direct_sum
 
+namespace module
+
+open_locale classical
+
+/-- A finitely generated torsion `ℤ`-module is finite. -/
+noncomputable def fintype_of_fg_torsion (G : Type*) [add_comm_group G] [module ℤ G]
+  [module.finite ℤ G] (hit : module.is_torsion ℤ G) : fintype G :=
+let hG := (module.equiv_direct_sum_of_is_torsion hit).some_spec in
+@fintype.of_equiv _ _
+  (@fintype.of_equiv _ _
+    (@pi.fintype _ _ _ hG.some $ λ i,
+      @fintype.of_equiv _ _
+        (@zmod.fintype _
+          ⟨int.nat_abs_pos_of_ne_zero $ pow_ne_zero (hG.some_spec.some_spec.some_spec.some i) $
+            irreducible.ne_zero $ hG.some_spec.some_spec.some i⟩)
+        (int.quotient_span_equiv_zmod _).symm.to_equiv)
+    (@dfinsupp.equiv_fun_on_fintype _ _ _ hG.some).symm)
+  hG.some_spec.some_spec.some_spec.some_spec.some.symm.to_equiv
```

Comment:
I'm wondering if this sort of approach might be better:
```suggestion
variables (G : Type*) [add_comm_group G] [module ℤ G] [module.finite ℤ G]

/-- A finitely generated torsion `ℤ`-module is finite. -/
lemma finite_of_fg_torsion (hit : module.is_torsion ℤ G) : _root_.finite G :=
begin
  obtain ⟨a, b, c, d, e, ⟨f⟩⟩ := module.equiv_direct_sum_of_is_torsion hit,
  haveI := λ i, fact.mk $ int.nat_abs_pos_of_ne_zero $ pow_ne_zero (e i) (d i).ne_zero,
  haveI : ∀ i, _root_.finite (ℤ ⧸ (submodule.span ℤ {c i ^ e i} : submodule ℤ ℤ)) :=
  λ i, finite.of_equiv _ (int.quotient_span_equiv_zmod (c i ^ e i)).symm.to_equiv,
  haveI : _root_.finite ⨁ i, ℤ ⧸ (submodule.span ℤ {c i ^ e i} : submodule ℤ ℤ) :=
  finite.of_equiv _ dfinsupp.equiv_fun_on_fintype.symm,
  exact finite.of_equiv _ f.symm.to_equiv,
end

noncomputable def fintype_of_fg_torsion (hit : module.is_torsion ℤ G) : fintype G :=
@fintype.of_finite G (finite_of_fg_torsion G hit)
```

Reply:
Or perhaps even without defining the `fintype` version at all.

Reply:
From what I was told, proving `finite` then applying `fintype.of_finite` is kind of pointless since the data is lost, so perhaps skipping `fintype` altogether is fine, but I'm not sure if we care about the data here. I defined the `fintype` directly since I literally can, and also because I'm more familiar with the `fintype` API.

Reply:
I've removed the `fintype` instances in favour of `finite` lemmas.